### ISSUE TITLE
resolve api supports skipSelf options

### ISSUE THREAD

**Issue by hardfist** [evanw/esbuild#2032]
I want to set most of  lib in node_modules as external, and try to use resolve hook to implement this
```ts
{ 
  name: 'external',
  setup(build){ 
    build.onResolve(({filter:/.*/}, args => {
        const realPath = build.Resolve(args.path, {resolveDir: args.resolveDir}); 
        if(realPath.includes('node_modules') && excludeSomeLib(realPath)){
               return { path: args.path, external: true}
        }else {
             return undefined;
        }
    })
  }
}
```
But it will cause infinite loop, rollup introduce [`skipSelf`](https://rollupjs.org/guide/en/#thisresolve) options to avoid this problem, could esbuild support this too?

**Comment by mhsdesign** [comment]
the above example might not be really performant but i ran now often into the same problem.

usually one needs to rebuild the `resolve` functionality -> which is currently basically impossible to get right. `require.resolve` doesnt work fx. with the `modules` option in the package json...

so the builtin resolve function is great and as soon as one builds a plugin with a non esbuild resolve functionality -> something will break at somepoint because esbuilds resolve is better ^^.
(i was debugging lately a few plugins because their resolve dindt got the right target ...)

so i would really like either a way to get an unrelated esbuild `resolve` function or to skipSelf.
my usecase;
```
{ 
  name: 'somePackageOnlyToBeBundledOnce',
  setup(build){ 
    build.onResolve(({filter:/^foo-bar(/.*)?/}, ({path}) => build.resolve(path, {resolveDir: __dirname}))
  }
}
```
(when linking a package: esbuild sometimes used `foo-bar` of the wrong node_modules -> and then its bundled twice)

**Comment by hardfist** [comment]
I agree with @mhsdesign , There are actually two different usage cases we need to use build.onResolve
* get the real resolved Id in resolve hooks and decide to whether make this id external | sideEffects
* use the internal esbuild's node resolution algorithm(without trigger esbuild's plugin), which is very hard to mimic in userland( I currently mimic it by enhanced-resolve, but it comes with inconsistency and hard to deal with the tsconfig's paths and baseUrl behavior)
So It would be helpful to support `skipSelf` and `skipPlugin` options to support these usage case.

**Comment by mhsdesign** [comment]
i think there is alreay an option - we just dindt see it ^^ namespaces.

this one would be for stubbing file './someFile' of module 'myTargetedPackage'

```js
import { sep } from 'path';
import { fileURLToPath } from 'url';
onResolve({ filter: /^\.\/someFile$/, namespace: 'file' }, ({ path, ...options }) => {
  if (options.importer.includes(`${sep}myTargetedPackage${sep}`)) {
    return resolve('./src/stubs/someFile.js', {
      resolveDir: fileURLToPath(new URL('.', import.meta.url)),
      namespace: 'myPluginNoRecurse',
    });
  }
  return resolve(path, {
    ...options,
    namespace: 'myPluginNoRecurse',
  });
});
```

**Comment by hardfist** [comment]
@mhsdesign I did try using namespace, but it also not works well, if other plugin use onResolve to handle all namespace

**Comment by mhsdesign** [comment]
can you explain this further? do you not want to limit you plugin to only the `file` namespaces but rather exclude the `myPluginNoRecurse` namespace?

i could imagine using regex for this too - or some simple negate flag like:

```js
onResolve({ filter: /^\.\/someFile$/, namespace: '!myPluginNoRecurse' }, ({ path, ...options }) => {
```

**Comment by aleclarson** [comment]
Related #3198
